query-languages/m/m-spec-basic-concepts.md

```diff
@@ -228,7 +228,7 @@ It does, however, rely on being able to reorder computations. Since expressions
 
 Once a value has been calculated, it is _immutable_, meaning it can no longer be changed. This simplifies the model for evaluating an expression and makes it easier to reason about the result since it is not possible to change a value once it has been used to evaluate a subsequent part of the expression. For instance, a record field is only computed when needed. However, once computed, it remains fixed for the lifetime of the record. Even if the attempt to compute the field raised an error, that same error will be raised again on every attempt to access that record field.
 
-An important exception to the immutable-once-calculated rule applies to list and table values. Both have _streaming semantics_. That is, repeated enumeration of the items in a list or the rows in a table can produce varying results. Streaming semantics enables the construction of M expressions that transform data sets that would not fit in memory at once.
+An important exception to the immutable-once-calculated rule applies to list, table and binary values, each of which has _streaming semantics_. M's representation of a list, table or binary value, while immutable, does not contain materialized list items, table rows or binary bytes (respectively). Instead, these are produced as needed each time the value is enumerated. Since a streamed value's items, rows or bytes are produced on demand, repeated enumerations of the value can produce varying results unless the source and all intermediate transforms are guaranteed to be deterministic. Streaming semantics enables the construction of M expressions that transform data sets that would not fit in memory at once.
```

Comment:
```suggestion
An important exception to the immutable-once-calculated rule applies to list, table, and binary values, which have _streaming semantics_. Streaming semantics allow M to transform data sets that do not fit into memory all at once. With streaming, the values returned when enumerating a given table, list, or binary value are produced as needed each time they're requested. Instead of being fixed and immutable, the enumerated values can therefore be dynamic and produce different results across multiple enumerations. This doesn't mean that multiple enumerations will always be different, just that they can sometimes be different, depending on whether the M being used always returns the same results.
```

Reply:
@ehrenMSFT, what do you think of this edit? It is based on what you wrote, but tweaked + expanded to indicate that both non-determinism in M code _and in the data source_ can result in different values coming out of different enumerations.

Reply:
I think it's good to mention that the data source can cause the dynamism.

This line isn't accurate, though:
_Since the enumerated values are not fixed and immutable but instead are produced dynamically on demand_

It should say that enumerated values _can_ be or _might_ be dynamic. If I declare a list {0..10}, there's no dynamism, even when streaming it.

Reply:
Thanks for the feedback, @ehrenMSFT! Is the latest commit any better?